packages/core/contracts/financial-templates/funding-rate-store/implementation/FundingRateStore.sol

```diff
@@ -48,51 +49,69 @@ contract FundingRateStore is FundingRateStoreInterface, Testable, Lockable {
 
     enum ProposalState { None, Pending, Expired }
 
-    mapping(bytes32 => FundingRateRecord) public fundingRateRecords;
-    mapping(bytes32 => mapping(uint256 => FundingRateRecord)) public fundingRateDisputes;
+    mapping(address => FundingRateRecord) public fundingRateRecords;
+    mapping(address => mapping(uint256 => FundingRateRecord)) public fundingRateDisputes;
 
     uint256 public proposalLiveness;
 
+    FixedPoint.Unsigned public fixedFundingRateFeePerSecondPerPfc; // Percentage of 1 E.g., .1 is 10% FundingRate fee.
+    FixedPoint.Unsigned public weeklyDelayFeePerSecondPerPfc; // Percentage of 1 E.g., .1 is 10% weekly delay fee.
+    uint256 public constant SECONDS_PER_WEEK = 604800;
     /****************************************
      *                EVENTS                *
      ****************************************/
 
-    event ProposedRate(bytes32 indexed identifier, int256 rate, uint256 proposalTime, address indexed proposer);
+    event ProposedRate(address indexed perpetual, int256 rate, uint256 proposalTime, address indexed proposer);
     event DisputedRate(
-        bytes32 indexed identifier,
+        address indexed perpetual,
         int256 rate,
         uint256 proposalTime,
         address indexed proposer,
         address indexed disputer
     );
-    event PublishedRate(bytes32 indexed identifier, int256 rate, uint256 proposalTime, address indexed proposer);
-    event FinalFeesPaid(uint256 indexed amount);
+    event PublishedRate(address indexed perpetual, int256 rate, uint256 proposalTime, address indexed proposer);
+    event FinalFeesPaid(address indexed collateralCurrency, uint256 indexed amount);
 
     constructor(
+        FixedPoint.Unsigned memory _fixedFundingRateFeePerSecondPerPfc,
+        FixedPoint.Unsigned memory _weeklyDelayFeePerSecondPerPfc,
         uint256 _proposalLiveness,
-        address _collateralAddress,
         address _finderAddress,
         address _timerAddress
     ) public Testable(_timerAddress) {
+        // TODO: Should we make the Perpetual contract the withdrawer of this Store?
+        _createExclusiveRole(uint256(Roles.Owner), uint256(Roles.Owner), msg.sender);
+        _createWithdrawRole(uint256(Roles.Withdrawer), uint256(Roles.Owner), msg.sender);
+
         require(_proposalLiveness > 0, "Proposal liveness is 0");
         proposalLiveness = _proposalLiveness;
-        collateralCurrency = IERC20(_collateralAddress);
         finder = FinderInterface(_finderAddress);
+
+        // Continuous fees at or over 100% don't make sense.
+        require(_fixedFundingRateFeePerSecondPerPfc.isLessThan(1), "Fee must be < 100% per second.");
+        require(_weeklyDelayFeePerSecondPerPfc.isLessThan(1), "weekly delay fee must be < 100%");
+
+        // TODO: Should we make these fee rates modifiable (by some admin) post deployment?
```

Comment:
Open questions about funding rate fees:
- Should we have a late penalty for this? I included because we want to enforce that fees are continuously paid, mostly taking inspiration from the Regular fee logic
- Should we allow these fee rates to be changeable post-deployment? If so, by whom?
- Should we allow a way for the Perpetual to withdraw fees?

Reply:
1. Seems reasonable, but don't have a strong opinion.
2. Maybe a similarly relevant question is: where should the fees be set -- in the funding rate store as a global value for all contracts? Or should they be set per-contract as a deployment param? In the former case, it seems like it would make sense to make it changeable by the DVM since it's a system-wide param. In the latter case, it seems more like a contract param that you might want to be immutable. I would lean towards the latter, but curious to hear others' thoughts.
3. The only case I can think of where this might be necessary would be a shutdown of some type that terminates the contract or a funding rate store migration. I could see us sending the fees back to the contract in those cases, but it sounds complicated to handle that. I wonder if there's a simpler way to deal with one or both of those cases. Any ideas?

Reply:
2.
> "it's a system-wide param"
My only counter to this is that this is not a DVM system param, it is a "uLabs" or Risk labs system param. I also lean towards the latter and make it immutable per each contract. Therefore the change to make is to store `_fixedFundingRateFeePerSecondPerPfc` and  `_weeklyDelayFeePerSecondPerPfc` in the Perpetual and not the Store

3. Agreed that we don't want to have a situation where we decide to shutdown the Perpetual/Store and have collateral frozen in the Store. We could relinquish these on an emergency shutdown?